Support opc ua certificate

Falco20019 opened the pull request:
Fixes #234
Fixes #237

Needs a bit more testing, just so that you can have a first look.

nauful commented:
One more rebase please, I just updated serialization and deserialization for token lifetimes.

nauful commented:
I'm getting weird errors for decoding token lifetime, it looks like uaexpert expects 4 byte widths. I'm going to revert these lifetime changes for now and revisit later.
<img width="1543" height="1191" alt="image" src="https://github.com/user-attachments/assets/e6ebbc02-40e0-4f64-9266-4a153b84b5b3" />
<img width="1192" height="540" alt="image" src="https://github.com/user-attachments/assets/d78e0a92-aaf8-4170-ae43-e6fc15369708" />

Falco20019 commented:
@nauful OK, I also reverted the double back to uint on the lifetime. Still strange that it's not matching the specification. Maybe it's only uint in this location and double on some others?

I also added unit tests and adjusted the server fixture to allow for explicit token checking.

Falco20019 commented:
.NET 6 and .NET 7 are EOL by the way. Starting with .NET 8 there are a lot better X.509 capabilities in the framework (especially regarding DER + PEM support).

I would like to give the library some love and do some refactorings to ensure everything matches the specs (especially extracting some more data types based on https://reference.opcfoundation.org/Core/Part4/v105/docs/7 like `SignatureData`). Also some cleanup as there is quite a lot copy-paste ongoing.

If you would be willing to follow the [semantic versioning standard](https://semver.org/) for the library, then I would create it as `2.0.0` since dropping target frameworks would be a breaking change and we could have it start with .NET 10 (upcoming LTS, releasing next week). Only dropping .NET 6+7 would also be possible, but .NET 8 still needs some BCL packages. So if you want to keep 8+9 (as they are still supported for 1 more year), I could also factor that in (meaning you will have another major version November 2026 when dropping 8+9).

I will not do any refactoring for .NET Framework as you already mentioned it's considered legacy. If you are fine with all of this, I might have a look as soon as this PR is merged :) Will still try to get some more testing against real servers done today.

Falco20019 commented:
I was able to validate that it's working against `docker run --rm --name opc-plc -p 7718:50000 -v $(pwd)/pki:/app/pki mcr.microsoft.com/iotedge/opc-plc:2.12.44 --cdn opc-plc --ph opc-plc`. I adjusted the `TestClient` a bit to export two different certificates to ensure there is no mixups happening. If you put the generated certificates into the trusted folders, the connection will work and session is activated. The `own` folder will be generated, so it's enough have the `trusted/certs` and `trusted-user/certs`. If you are not doing it upfront, they will just end up in `rejected/certs` and you could move them and reconnect. But doing it upfront ensures it's REALLY the ones we expect to send :)

<img width="216" height="200" alt="image" src="https://github.com/user-attachments/assets/a1ab6965-e0c9-442b-b249-bc8ff84cabef" />

Falco20019 commented:
@nauful Hey, any ETA on when we can look into getting this merged? Thanks for your support.

nauful commented:
Re: https://github.com/nauful/LibUA/pull/238#issuecomment-3495374215 Yes please, I do appreciate all the effort! I haven't done anything with UA in a while, so all of this helps.

One last thing, could you please check that `TestServer` can receive a certificate based user and verify, similar to how it shows you how to verify anonymous or user/pass? I saw the tests and they look good. You might do this in a later PR, I will merge now so you aren't blocked.

nauful commented:
Hmm, the branch tests were file but encryption fails after merging. Could you please also test on Windows? I tried connecting from TestClient to TestServer.

```
Internal.Cryptography.CryptoThrowHelper.WindowsCryptographicException: 'The parameter is incorrect.'
```

```
>	System.Security.Cryptography.Algorithms.dll!System.Security.Cryptography.RSAImplementation.RSACng.EncryptOrDecrypt(Microsoft.Win32.SafeHandles.SafeNCryptKeyHandle key, System.ReadOnlySpan<byte> input, Interop.NCrypt.AsymmetricPaddingMode paddingMode, void* paddingInfo, bool encrypt) Line 340	C#
 	System.Security.Cryptography.Algorithms.dll!System.Security.Cryptography.RSAImplementation.RSACng.EncryptOrDecrypt(byte[] data, System.Security.Cryptography.RSAEncryptionPadding padding, bool encrypt) Line 221	C#
 	System.Security.Cryptography.Algorithms.dll!System.Security.Cryptography.RSAImplementation.RSACng.Decrypt(byte[] data, System.Security.Cryptography.RSAEncryptionPadding padding) Line 146	C#
 	LibUA.dll!LibUA.UASecurity.Decrypt(System.ArraySegment<byte> data, System.Security.Cryptography.X509Certificates.X509Certificate2 cert, System.Security.Cryptography.RSA rsaPrivate, System.Security.Cryptography.RSAEncryptionPadding padding) Line 849	C#
 	LibUA.dll!LibUA.Client.OpenSecureChannelInternal(bool renew) Line 353	C#
 	LibUA.dll!LibUA.Client.OpenSecureChannel(LibUA.Core.MessageSecurityMode messageSecurityMode, LibUA.Core.SecurityPolicy securityPolicy, byte[] serverCert) Line 114	C#
 	TestClient.dll!TestClient.Program.Main(string[] args) Line 166	C#
```

nauful commented:
Nevermind, it looks like something is wrong on my system even with the old version. I'll check it again when I have time, I'll put up the new nuget package for now.
